helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -775,12 +840,35 @@ private void handleEvent(ClusterEvent event, BaseControllerDataProvider dataProv
 
     _helixManager = manager;
 
-    // TODO If init controller with paused = true, it may not take effect immediately
-    // _paused is default false. If any events come before controllerChangeEvent, the controller
-    // will be excuting in un-paused mode. Which might not be the config in ZK.
-    if (_paused) {
-      logger.info("Cluster " + manager.getClusterName() + " is paused. Ignoring the event:" + event
-          .getEventType());
+    // Prepare ClusterEvent
+    // TODO (harry): this is a temporal workaround - after controller is separated we should not
+    // have this instanceof clauses
+    List<Pipeline> pipelines;
+    boolean isTaskFrameworkPipeline = false;
+    Pipeline.Type pipelineType;
+
+    if (dataProvider instanceof ResourceControllerDataProvider) {
+      pipelines = _registry.getPipelinesForEvent(event.getEventType());
+      pipelineType = Pipeline.Type.DEFAULT;
+    } else if (dataProvider instanceof WorkflowControllerDataProvider) {
+      pipelines = _taskRegistry.getPipelinesForEvent(event.getEventType());
+      isTaskFrameworkPipeline = true;
+      pipelineType = Pipeline.Type.TASK;
+    } else if (dataProvider instanceof ManagementControllerDataProvider) {
+      pipelines = _managementModeRegistry.getPipelinesForEvent(event.getEventType());
+      pipelineType = Pipeline.Type.MANAGEMENT_MODE;
+    } else {
+      logger.warn(String
+          .format("No %s pipeline to run for event: %s::%s", dataProvider.getPipelineName(),
+              event.getEventType(), event.getEventId()));
+      return;
+    }
+
+    // Default/task pipelines should not run while management mode pipeline is enabled to run
+    if (!Pipeline.Type.MANAGEMENT_MODE.equals(pipelineType)
+        && _managementControllerDataProvider.shouldRunManagementPipeline()) {
```

Comment:
This is not a good idea to take the cache for management pipelines. It can cause race conditions.

I would suggest to use the cache from the event since this function is general function for all three types of caches.

Reply:
This is a very tricky place. The method `shouldRunManagementPipeline()` should act differently for the default and management pipelines (on the contrary)
 
I tried to use the dedicated cache. It's really difficult to switch and update the cache correctly: when management pipeline is running, it has the new data. But for the default pipeline cache, it still has the old data. The default pipeline's old data won't work. It'd be really difficult to update the data or signal the refresh from management pipeline.

 Thus, it'd be better and easier to have a single source of truth, which is the management cache. I'll sync with you offline for more details.